File: refactored/assertion-roulette/python-pytest/ApplyPagination_Before_Last_10.py
import pytest

class TestCursorPagingHandler:

    def test_apply_pagination_before_last_10(self):
        # arrange
        data = [
            Foo.create(0),
            Foo.create(1),
            Foo.create(2),
            Foo.create(3),
        ]
        # act
        result = apply(data, to_base64(3), 10)
        # assert
        assert to_foo(result)[0].get_index() == 0, "Explanation message"
        assert to_foo(result)[-1].get_index() == 2, "Explanation message"
        assert result.get_info().has_next_page(), "Explanation message"
        assert not result.get_info().has_previous_page(), "Explanation message"
        assert result.get_info().get_start_cursor() == to_base64(0), "Explanation message"
        assert result.get_info().get_end_cursor() == to_base64(2), "Explanation message"
        assert result.get_total_count_async() == 4, "Explanation message"

File: assertion-roulette/java-junit/ApplyPagination_Before_Last_10.java
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

class CursorPagingHandlerTests {

    @Test
    void applyPagination_Before_Last_10() {
        // arrange
        Foo[] data = {
            Foo.create(0),
            Foo.create(1),
            Foo.create(2),
            Foo.create(3),
        };
        // act
        var result = await apply(data, before: toBase64(3), last: 10);
        // assert
        assertEquals(0, toFoo(result).get(0).getIndex());
        assertEquals(2, toFoo(result).get(toFoo(result).size() - 1).getIndex());
        assertTrue(result.getInfo().hasNextPage());
        assertFalse(result.getInfo().hasPreviousPage());
        assertEquals(toBase64(0), result.getInfo().getStartCursor());
        assertEquals(toBase64(2), result.getInfo().getEndCursor());
        assertEquals(4, await result.getTotalCountAsync(default));
    }
}
